### PR TITLE
Add recursive deletion of components for `zenml stack delete`

### PR DESCRIPTION
**Pull request by KenmogneThimotee**
## Describe changes
I implemented/fixed _ to achieve _.

## Pre-requisites
Please ensure you have done the following:
- [* ] I have read the **CONTRIBUTING.md** document.
- [* ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [* ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [* ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Comment by schustmi** [comment]
Hey @KenmogneThimotee, thanks for the PR! The implementation looks good to me, however I would ask you to move the logic regarding the stack component deletion into the `Client.delete_stack(...)` method instead.

Some explanation on why: The CLI is only one of the ways to interact with ZenML, and some people prefer to instead use our `Client` class directly in code. For that reason we try to implement logic such as recursive deletion etc. in the client methods instead.

**Comment by KenmogneThimotee** [comment]
ok, I will do that. Thanks.

**Comment by KenmogneThimotee** [comment]
Do I need to keep the cli code because it informs the user of what is going to happen when it chooses to delete the stack recursively ?

**Comment by schustmi** [comment]
Yes, I'd suggest keeping the confirmation code for the recursive delete inside the CLI, and if the user confirmed call `Client().delete_stack(..., recursive=True)` which handles the actual recursive deletion.

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/client.py [hunk 1]
```diff
@@ -1779,6 +1780,46 @@ def delete_stack(self, name_id_or_prefix: Union[str, UUID]) -> None:
...
+        if recursive:
+
+    
+            stack_components_free_for_deletion = []
+            stack_components_not_free_for_deletion = []
+
+            # Get all stack components associated with this stack
+            stack_components = stack.components
+
+            for stack_component in stack_components.items():
```
[the remark this answers is not in the document]
Comment (KenmogneThimotee) [reply src/zenml/client.py:1792]: it makes more sense.

File: src/zenml/client.py [hunk 2]
```diff
@@ -1779,6 +1780,46 @@ def delete_stack(self, name_id_or_prefix: Union[str, UUID]) -> None:
                 f"one."
             )
 
+        if recursive:
+
+    
+            stack_components_free_for_deletion = []
+            stack_components_not_free_for_deletion = []
```
[the remark this answers is not in the document]
Comment (KenmogneThimotee) [reply src/zenml/client.py:1787]: The second array is not necessary but the first yes because I need to keep all the components that can be removed

**Review by KenmogneThimotee — COMMENTED** [review]
I added all review

File: src/zenml/client.py [hunk 3]
```diff
@@ -1779,6 +1782,44 @@ def delete_stack(self, name_id_or_prefix: Union[str, UUID]) -> None:
                 f"one."
             )
 
+        if recursive:
+
```
Comment (htahir1) [inline src/zenml/client.py:1786]:
```suggestion
```

**Review by htahir1 — CHANGES_REQUESTED** [review]
Minor nits from my side. Otherwise LGTM!

File: src/zenml/client.py [hunk 4]
```diff
@@ -1779,6 +1782,44 @@ def delete_stack(self, name_id_or_prefix: Union[str, UUID]) -> None:
...
+            self.delete_stack(stack.id)
+
+            for (
+                stack_component_type,
+                stack_component_model,
+            ) in stack_components_free_for_deletion:
+                self.delete_stack_component(
+                    stack_component_model[0].name, stack_component_type
+                )
+                logger.info(
```
Comment (schustmi) [inline src/zenml/client.py:1815]: I think this log is already happening in `self.delete_stack_component(...)` right? If that's the case, I would remove it here.

**Review by schustmi — CHANGES_REQUESTED** [review]

**Comment by schustmi** [comment]
Awesome, thank you very much for the contribution!

**Pull request merged**